<style  type="text/css"> 
</style>

<b><center>
<span style="font-size: 24pt; line-height: 1.2">
COMS W4111-002 (Fall 2021)<br>Introduction to Databases
</span>
</center></b>
</span><br>
<p>
<i><center>
<span style="font-size: 20pt; line-height: 1.2">
Homework 1: Programming - 10 Points
</span>
</center></i>

__Note:__ Please replace the information below with your last name, first name and UNI.<br><br>



<i>
<span style="font-size: 20pt; line-height: 1.2"; >
Huo_Yanyi, yh3420
</span>
</i>

## Introduction

### Objectives
This homework has you practice and build skill with:

- PART A: (1 point) Understanding relational databases
- PART B: (1 point) Understanding relational algebra
- PART C: (1 point) Cleaning data
- PART D: (1 point) Performing simple SQL queries to analyze the data.
- PART E: (6 points) CSVDataTable.py


__Note:__ The motivation for PART E may not be clear. The motivation will become clearer as the semester proceeds. The purpose of PART E is to get you started on programming in Python and manipulating data.



### Submission
1. File > Print Preview > Download as PDF
2. Upload .pdf and .ipynb to GradeScope
3. Upload CSVDataTable.py and CSVDataTable_Tests.py

**This assignment is due September 24, 11:59 pm ET**


### Collaboration
- You may use any information you get in TA or Prof. Ferguson's office hours, from lectures or from recitations.
- You may use information that you find on the web.
- You are NOT allowed to collaborate with other students outside of office hours.


# Part A: Written

1. What is a database management system?

    <i> A database management system is a software package designed to define, manipulate, retrieve and manage data in a database. (source: website) </i>

2. What is a primary key and why is it important?

    <i> Primary key is a value, or a combination of few values from the table, uniquely defining each record in this table. It can uniqely identifies each row in a database table. (source: website)</i>

3. Please explain the differences between SQL, MySQL Server and DataGrip?

    <i> SQL is a query language, whereas MySQL is a relational database that uses SQL to query a database. You can use SQL to access, update, and manipulate the data stored in a database. However, MySQL is a database that stores the existing data in a database in an organized manner. DataGrip is a cross-platform IDE that is aimed at DBAs and developers working with SQL databases. (source: website)</i>

4. What are 4 different types of DBMS table relationships, give a brief explanaition for each?

    <i> one to one: each record of one table is related to only one record of the other table</i>
    
    one to many: each record of one table can be related to one or more than one record of the other table</i>
    
    many to one: a single record of the second table can be related to one or more than one record of the first table
     
     many to many: each record of the first table can be related to one or more than one record of the second table and a single record of the second table can be related to one or more than one record of the first table (source: website)</i>

5. What is an ER model? 

    <i> Entity Model is a graphical approach to database design. It is a high-level data model that defines data elements and their relationship for a specified software system. (source: website)</i>


6. Using Lucidchart draw an example of a logical ER model using Crow's Foot notation for Columbia classes. The entity types are:
    - Students, Professors, and Classes.
    - The relationships are:
        - A Class has exactly one Professor.
        - A Student has exactly one professor who is an _advisor._
        - A Professor may advise 0, 1 or many Students.
        - A Class has 0, 1 or many enrolled students.
        - A Student enrolls in 0, 1 or many Classes.
        
        
- You can define what you think are common attributes for each of the entity types. Do not define more than 5 or 6 attributes per entity type.


- In this example, explicitly show an example of a primary-key, foreign key, one-to-many relationship, and many-to-many relationship.


__Notes:__
- If you have not already done so, please register for a free account at Lucidchart.com. You can choose the option at the bottom of the left pane to add the ER diagram shapes.
- You can take a screen capture of you diagram and save in the zip directory that that contains you Jupyter notebook. Edit the following cell and replace "Boromir.jpg" with the name of the file containing your screenshot.
    
    

   <i> Use the following line to upload a photo of your Luicdchart. <i/>
   <img src="lucidchart.png">

# Part B: Relational Algebra

You will use [the online relational calculator](https://dbis-uibk.github.io/relax/landing), choose the “Karlsruhe University of Applied Sciences” dataset.

An anti-join is a form of join with reverse logic. Instead of returning rows when there is a match (according to the join predicate) between the left and right side, an anti-join returns those rows from the left side of the predicate for which there is no match on the right. 

The Anti-Join Symbol is ▷.

Consider the following relational algebra expression and result.

/* (1) Set X = The set of classrooms in buildings Taylor or Watson. */

        X = σ building='Watson' ∨ building='Taylor' (classroom)

/* (2) Set Y =  The Anti-Join of department and X */

        Y = (department ▷ X)

/* (3) Display the rows in Y. */

        Y
        
   <img src="ra.png">


1. Find an alternate expression to (2) that computes the correct answer given X. Display the execution of your query below.

    <img src="hw1_B.png">
    <img src="hw1_B_2.png">


# Part C: Data Clean Up

## Please note: You MUST make a new schema using the lahmansdb_to_clean.sql file provided in the data folder. 

Use thelahmansdb_to_clean.sql file to make a new schema containing the raw data. The lahman database you created in Homework 0 has already been cleaned with all the constraints and will be used for Part D. Knowing how to clean data and add integrity constraints is very important which is why you go through the steps in part C. 

TLDR: If you use the HW0 lahman schema for this part you will get a lot of errors and recieve a lot of deductions. 

In [1]:
# You will need to follow instructions from HW 0 to make a new schema, import the data. 
# Connect to the unclean schema below by setting the database host, user ID and password.
%load_ext sql
%sql mysql+pymysql://root:19990105Hyy@localhost/lahmansdb_to_clean


Data cleanup: For each table we want you to clean, we have provided a list of changes you have to make. You can reference the cleaned lahman db for inspiration and guidance, but know that there are different ways to clean the data and you will be graded for your choice rationalization. You should make these changes through DataGrip's workbench's table editor and/or using SQL queries. In this part you will clean two tables: People and Batting. 


### You must have: 
- A brief explanation of why you think the change we requested is important.
- What change you made to the table.
- Any queries you used to make the changes, either the ones you wrote or the Alter statements provided by DataGrip's editor. 
- Executed the test statements we provided
- The cleaned table's new create statement (after you finish all the changes) 


### Overview of Changes:

People Table 

   0. Primary Key (Explanation is given, but you still must add the key to your table yourself)
   1. Empty strings to NULLs
   2. Column typing
   3. isDead column
   4. deathDate and birthDate column

Batting Table

   1. Empty strings to NULLs
   2. Column typing
   3. Primary Key
   4. Foreign Key
   

   
### How to make the changes: 

__Using the Table Editor:__

When you hit apply, a popup will open displaying the ALTER statments sql generates. Copy the sql provided first and paste it into this notebook. Then you can apply the changes. This means that you are NOT executing the ALTER statements through your notebook. 


1. Right click on the table > Modify Table...
<img src="modify.png" width="400" height="800">

2. Keys > press the + button > input the parameters > Execute 
    OR Keys > press the + button > input the parameters > copy and paste the script generated under "SQL Script" and paste into your notebook > Run the cell in jupyter notebook
<img src="pk.png" width="600" height="1200">




__Using sql queries:__

Copy paste any queries that you write manually into the notebook as well!


<hr style="height:2px">

## People Table

### 0) EXAMPLE: Add a Primary Key 

(Solutions are given but make sure you still do this step in workbench!)

#### Explanation

We want to add a Primary Key because we want to be able to uniquely identify rows within our data. A primary key is also an index, which allows us to locate data faster. 

#### Change

I added a Primary Key on the playerID column and made the datatype VARCHAR(15)

__Note:__ This is for demonstration purposes only. playerID __is not__ a primary key for fielding.

#### SQL

~~~~sql
ALTER TABLE `lahmansdb_to_clean`.`people`
CHANGE COLUMN `playerID` `playerID` VARCHAR(15) NOT NULL ,
ADD PRIMARY KEY (`playerID`);
~~~~

#### Tests

In [2]:
%sql SHOW KEYS FROM people WHERE Key_name = 'PRIMARY'
    

 * mysql+pymysql://root:***@localhost/lahmansdb_to_clean
1 rows affected.


[('people', 0, 'PRIMARY', 1, 'playerID', 'A', 20071, None, None, '', 'BTREE', '', '', 'YES', None)]

### 1) Convert all empty strings to NULL

#### Explanation

_The type of empty string is only string but NULL type can be anyother types._

#### Change

_I update each column to NULL when it is an empty string in the people table._

#### SQL

~~~~sql
update people set birthYear = NULL where birthYear = '';
update people set birthMonth = NULL where birthMonth = '';
update people set birthDay = NULL where birthDay = '';
update people set birthCountry = NULL where birthCountry = '';
update people set birthState = NULL where birthState = '';
update people set birthCity = NULL where birthCity = '';
update people set deathYear = NULL where deathYear = '';
update people set deathMonth = NULL where deathMonth = '';
update people set deathDay = NULL where deathDay = '';
update people set deathCountry = NULL where deathCountry = '';
update people set deathState = NULL where deathState = '';
update people set deathCity = NULL where deathCity = '';
update people set nameFirst = NULL where nameFirst = '';
update people set nameLast = NULL where nameLast = '';
update people set nameGiven = NULL where nameGiven = '';
update people set weight = NULL where weight = '';
update people set height = NULL where height = '';
update people set bats = NULL where bats = '';
update people set throws = NULL where throws = '';
update people set debut = NULL where debut = '';
update people set finalGame = NULL where finalGame = '';
update people set retroID = NULL where retroID = '';
update people set bbrefID = NULL where bbrefID = '';
~~~~

#### Tests

In [3]:
%sql SELECT * FROM people WHERE birthState = ""


 * mysql+pymysql://root:***@localhost/lahmansdb_to_clean
0 rows affected.


[]

### 2) Change column datatypes to appropriate values (ENUM, INT, VARCHAR, DATETIME, ETC)

#### Explanation

_I can easily operate the data in the following step with appropriate datatypes._

#### Change

_I change datatypes from text to appropriate values._

#### SQL

~~~~sql
alter table people modify birthYear numeric(4) null;
alter table people modify birthMonth numeric(2) null;
alter table people modify birthDay numeric(2) null;
alter table people modify birthCountry varchar(20) null;
alter table people modify birthState varchar(25) null;
alter table people modify birthCity varchar(30) null;
alter table people modify deathYear numeric(4) null;
alter table people modify deathMonth numeric(2) null;
alter table people modify deathDay numeric(2) null;
alter table people modify deathCountry varchar(25) null;
alter table people modify deathState varchar(25) null;
alter table people modify deathCity varchar(30) null;
alter table people modify nameFirst varchar(20) null;
alter table people modify nameLast varchar(20) null;
alter table people modify nameGiven varchar(50) null;
alter table people modify weight int null;
alter table people modify height int null;
alter table people modify bats char null;
alter table people modify throws char null;
alter table people modify debut date null;
alter table people modify finalGame date null;
alter table people modify retroID varchar(15) null;
alter table people modify bbrefID varchar(15) null;
~~~~

### 3) Add an isDead Column that is either 'Y' or 'N'
- Some things to think of: What data type should this column be? How do you know if the player is dead or not? Maybe you do not know if the player is dead.
- You do not need to make guesses about life spans, etc. Just apply a simple rule.

'Y' means the player is dead

'N' means the player is alive

#### Explanation

_By looking at isDead, I can quickly know this person is alive or not._

#### Change

_I add an isDead column and set it to 'N' when the deathYear is null. Then I set other null value in isDead to 'Y'. _

#### SQL

~~~~sql
alter table people add isDead char null;
update people set isDead = 'N' where deathYear IS NULL;
update people set isDead = 'Y' where isDead IS NULL;
~~~~

#### Tests

In [4]:
%sql SELECT * FROM people WHERE isDead = "N" limit 10

 * mysql+pymysql://root:***@localhost/lahmansdb_to_clean
10 rows affected.


[('aardsda01', Decimal('1981'), Decimal('12'), Decimal('27'), 'USA', 'CO', 'Denver', None, None, None, None, None, None, 'David', 'Aardsma', 'David Allan', 215, 75, 'R', 'R', datetime.date(2004, 4, 6), datetime.date(2015, 8, 23), 'aardd001', 'aardsda01', 'N', None, datetime.date(1981, 12, 27)),
 ('aaronha01', Decimal('1934'), Decimal('2'), Decimal('5'), 'USA', 'AL', 'Mobile', None, None, None, None, None, None, 'Hank', 'Aaron', 'Henry Louis', 180, 72, 'R', 'R', datetime.date(1954, 4, 13), datetime.date(1976, 10, 3), 'aaroh101', 'aaronha01', 'N', None, datetime.date(1934, 2, 5)),
 ('aasedo01', Decimal('1954'), Decimal('9'), Decimal('8'), 'USA', 'CA', 'Orange', None, None, None, None, None, None, 'Don', 'Aase', 'Donald William', 190, 75, 'R', 'R', datetime.date(1977, 7, 26), datetime.date(1990, 10, 3), 'aased001', 'aasedo01', 'N', None, datetime.date(1954, 9, 8)),
 ('abadan01', Decimal('1972'), Decimal('8'), Decimal('25'), 'USA', 'FL', 'Palm Beach', None, None, None, None, None, None, 'Andy', 'Abad', 'Fausto Andres', 184, 73, 'L', 'L', datetime.date(2001, 9, 10), datetime.date(2006, 4, 13), 'abada001', 'abadan01', 'N', None, datetime.date(1972, 8, 25)),
 ('abadfe01', Decimal('1985'), Decimal('12'), Decimal('17'), 'D.R.', 'La Romana', 'La Romana', None, None, None, None, None, None, 'Fernando', 'Abad', 'Fernando Antonio', 235, 74, 'L', 'L', datetime.date(2010, 7, 28), datetime.date(2019, 9, 28), 'abadf001', 'abadfe01', 'N', None, datetime.date(1985, 12, 17)),
 ('abbotgl01', Decimal('1951'), Decimal('2'), Decimal('16'), 'USA', 'AR', 'Little Rock', None, None, None, None, None, None, 'Glenn', 'Abbott', 'William Glenn', 200, 78, 'R', 'R', datetime.date(1973, 7, 29), datetime.date(1984, 8, 8), 'abbog001', 'abbotgl01', 'N', None, datetime.date(1951, 2, 16)),
 ('abbotje01', Decimal('1972'), Decimal('8'), Decimal('17'), 'USA', 'GA', 'Atlanta', None, None, None, None, None, None, 'Jeff', 'Abbott', 'Jeffrey William', 190, 74, 'R', 'L', datetime.date(1997, 6, 10), datetime.date(2001, 9, 29), 'abboj002', 'abbotje01', 'N', None, datetime.date(1972, 8, 17)),
 ('abbotji01', Decimal('1967'), Decimal('9'), Decimal('19'), 'USA', 'MI', 'Flint', None, None, None, None, None, None, 'Jim', 'Abbott', 'James Anthony', 200, 75, 'L', 'L', datetime.date(1989, 4, 8), datetime.date(1999, 7, 21), 'abboj001', 'abbotji01', 'N', None, datetime.date(1967, 9, 19)),
 ('abbotku01', Decimal('1969'), Decimal('6'), Decimal('2'), 'USA', 'OH', 'Zanesville', None, None, None, None, None, None, 'Kurt', 'Abbott', 'Kurt Thomas', 180, 71, 'R', 'R', datetime.date(1993, 9, 7), datetime.date(2001, 4, 13), 'abbok002', 'abbotku01', 'N', None, datetime.date(1969, 6, 2)),
 ('abbotky01', Decimal('1968'), Decimal('2'), Decimal('18'), 'USA', 'MA', 'Newburyport', None, None, None, None, None, None, 'Kyle', 'Abbott', 'Lawrence Kyle', 200, 76, 'L', 'L', datetime.date(1991, 9, 10), datetime.date(1996, 8, 24), 'abbok001', 'abbotky01', 'N', None, datetime.date(1968, 2, 18))]

### 4) Add a deathDate and birthDate column
Some things to think of: What do you do if you are missing information? What datatype should this column be?

You have to create this column from other columns in the table.

#### Explanation

_By adding deathDate and birthDate, I can simply find the date instead of looking three columns._

#### Change

_I add deathDate column and birthDate column first. Then I concat year, month and day into a string. The last step is I convert the string to date. _

#### SQL

~~~~sql
alter table people add deathDate date null;
alter table people add birthDate date null;
update people set deathDate = STR_TO_DATE(concat(deathYear,'/',deathMonth, '/',deathDay), '%Y/%m/%d')
where deathYear IS NOT NULL;
update people set birthDate = STR_TO_DATE(concat(birthYear,'/',birthMonth, '/',birthDay), '%Y/%m/%d')
where birthYear IS NOT NULL;
~~~~

#### Tests

In [5]:
%sql SELECT deathDate FROM people WHERE deathDate >= '2005-01-01' ORDER BY deathDate ASC LIMIT 10;

 * mysql+pymysql://root:***@localhost/lahmansdb_to_clean
10 rows affected.


[(datetime.date(2005, 1, 4),),
 (datetime.date(2005, 1, 7),),
 (datetime.date(2005, 1, 9),),
 (datetime.date(2005, 1, 10),),
 (datetime.date(2005, 1, 21),),
 (datetime.date(2005, 1, 22),),
 (datetime.date(2005, 1, 31),),
 (datetime.date(2005, 2, 4),),
 (datetime.date(2005, 2, 8),),
 (datetime.date(2005, 2, 11),)]

In [6]:
%sql SELECT birthDate FROM people WHERE birthDate <= '1965-01-01' ORDER BY birthDate ASC LIMIT 10;


 * mysql+pymysql://root:***@localhost/lahmansdb_to_clean
10 rows affected.


[(datetime.date(1820, 4, 17),),
 (datetime.date(1824, 10, 5),),
 (datetime.date(1832, 9, 17),),
 (datetime.date(1832, 10, 23),),
 (datetime.date(1835, 1, 10),),
 (datetime.date(1836, 2, 29),),
 (datetime.date(1837, 12, 26),),
 (datetime.date(1838, 3, 10),),
 (datetime.date(1838, 7, 16),),
 (datetime.date(1838, 8, 27),)]

### Final CREATE Statement

To find the create statement:

- Right click on the table name in workbench
- Select 'Copy to Clipboard'
- Select 'Create Statement' 

The create statement will now be copied into your clipboard and can be pasted into the cell below. 

~~~~sql
CREATE TABLE `people` (
  `playerID` varchar(15) NOT NULL,
  `birthYear` decimal(4,0) DEFAULT NULL,
  `birthMonth` decimal(2,0) DEFAULT NULL,
  `birthDay` decimal(2,0) DEFAULT NULL,
  `birthCountry` varchar(20) DEFAULT NULL,
  `birthState` varchar(25) DEFAULT NULL,
  `birthCity` varchar(30) DEFAULT NULL,
  `deathYear` int DEFAULT NULL,
  `deathMonth` int DEFAULT NULL,
  `deathDay` int DEFAULT NULL,
  `deathCountry` varchar(25) DEFAULT NULL,
  `deathState` varchar(25) DEFAULT NULL,
  `deathCity` varchar(30) DEFAULT NULL,
  `nameFirst` varchar(20) DEFAULT NULL,
  `nameLast` varchar(20) DEFAULT NULL,
  `nameGiven` varchar(50) DEFAULT NULL,
  `weight` int DEFAULT NULL,
  `height` int DEFAULT NULL,
  `bats` char(1) DEFAULT NULL,
  `throws` char(1) DEFAULT NULL,
  `debut` date DEFAULT NULL,
  `finalGame` date DEFAULT NULL,
  `retroID` varchar(15) DEFAULT NULL,
  `bbrefID` varchar(15) DEFAULT NULL,
  `isDead` char(1) DEFAULT NULL,
  `deathDate` date DEFAULT NULL,
  `birthDate` date DEFAULT NULL,
  PRIMARY KEY (`playerID`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci
~~~~

<hr style="height:2px">

## Batting Table

### 1) Convert all empty strings to NULL

#### Explanation

_The type of empty string is only string but NULL type can be anyother types._

#### Change

_I update each column to NULL when it is an empty string in the batting table._

#### SQL

~~~~sql
update batting set playerID = NULL where playerID = '';
update batting set yearID = NULL where yearID = '';
update batting set stint = NULL where stint = '';
update batting set teamID = NULL where teamID = '';
update batting set lgID = NULL where lgID = '';
update batting set G = NULL where G = '';
update batting set AB = NULL where AB = '';
update batting set R = NULL where R = '';
update batting set H = NULL where H = '';
update batting set 2B = NULL where 2B = '';
update batting set 3B = NULL where 3B = '';
update batting set HR = NULL where HR = '';
update batting set RBI = NULL where RBI = '';
update batting set SB = NULL where SB = '';
update batting set CS = NULL where CS = '';
update batting set BB = NULL where BB = '';
update batting set SO = NULL where SO = '';
update batting set IBB = NULL where IBB = '';
update batting set HBP = NULL where HBP = '';
update batting set SH = NULL where SH = '';
update batting set SF = NULL where SF = '';
update batting set GIDP = NULL where GIDP = '';
~~~~

#### Tests

In [7]:
%sql SELECT count(*) FROM lahmansdb_to_clean.batting where RBI is NULL;

 * mysql+pymysql://root:***@localhost/lahmansdb_to_clean
1 rows affected.


[(756,)]

### 2) Change column datatypes to appropriate values (ENUM, INT, VARCHAR, DATETIME, ETC)

#### Explanation

_I can easily operate the data in the following step with appropriate datatypes._

#### Change

_I change datatypes from text to appropriate values._

#### SQL

~~~~sql
alter table batting modify playerID varchar(15) null;
alter table batting modify yearID int null;
alter table batting modify stint int null;
alter table batting modify teamID varchar(5) null;
alter table batting modify lgID varchar(5) null;
alter table batting modify G int null;
alter table batting modify AB int null;
alter table batting modify R int null;
alter table batting modify H int null;
alter table batting modify `2B` int null;
alter table batting modify `3B` int null;
alter table batting modify HR int null;
alter table batting modify RBI int null;
alter table batting modify SB int null;
alter table batting modify CS int null;
alter table batting modify BB int null;
alter table batting modify SO int null;
alter table batting modify IBB int null;
alter table batting modify HBP int null;
alter table batting modify SH int null;
alter table batting modify SF int null;
alter table batting modify GIDP int null;
~~~~

### 3) Add a Primary Key
Two options for the Primary Key:

- Composite Key: playerID, yearID, stint
- Covering Key (Index): playerID, yearID, stint, teamID


#### Choice

_Composite Key: playerID, yearID, stint_

#### Explanation


_We want to add a Primary Key because we want to be able to uniquely identify rows within our data. A primary key is also an index, which allows us to locate data faster._

#### SQL

~~~~sql
alter table batting
	add constraint batting_pk
		primary key (playerID, yearID, stint);
~~~~

#### Test

In [8]:
%sql SHOW KEYS FROM batting WHERE Key_name = 'PRIMARY' and Column_name = 'playerID'

 * mysql+pymysql://root:***@localhost/lahmansdb_to_clean
1 rows affected.


[('batting', 0, 'PRIMARY', 1, 'playerID', 'A', 19607, None, None, '', 'BTREE', '', '', 'YES', None)]

### 4) Add a foreign key on playerID between the People and Batting Tables
Note: Two people in the batting table do not exist in the people table. How should you handle this issue?

#### Explanation

_It creates data integrity._

#### Change

_I add a foreign key in batting table and related to people table._

#### SQL

~~~~sql
alter table batting
	add constraint batting_people_playerID_fk
		foreign key (playerID) references people (playerID);
~~~~

#### Tests

In [9]:
%sql Select playerID from batting where playerID not in (select playerID from people);

 * mysql+pymysql://root:***@localhost/lahmansdb_to_clean
0 rows affected.


[]

In [10]:
%sql SELECT nameFirst, nameLast FROM people WHERE playerID = 'collied01';

 * mysql+pymysql://root:***@localhost/lahmansdb_to_clean
1 rows affected.


[('Eddie', 'Collins')]

### Final CREATE Statement

To find the create statement:

- Right click on the table name in workbench
- Select 'Copy to Clipboard'
- Select 'Create Statement' 

The create statement will now be copied into your clipboard and can be pasted into the cell below. 

~~~~sql
CREATE TABLE `batting` (
  `playerID` varchar(15) NOT NULL,
  `yearID` int NOT NULL,
  `stint` int NOT NULL,
  `teamID` varchar(5) DEFAULT NULL,
  `lgID` varchar(5) DEFAULT NULL,
  `G` int DEFAULT NULL,
  `AB` int DEFAULT NULL,
  `R` int DEFAULT NULL,
  `H` int DEFAULT NULL,
  `2B` int DEFAULT NULL,
  `3B` int DEFAULT NULL,
  `HR` int DEFAULT NULL,
  `RBI` int DEFAULT NULL,
  `SB` int DEFAULT NULL,
  `CS` int DEFAULT NULL,
  `BB` int DEFAULT NULL,
  `SO` int DEFAULT NULL,
  `IBB` int DEFAULT NULL,
  `HBP` int DEFAULT NULL,
  `SH` int DEFAULT NULL,
  `SF` int DEFAULT NULL,
  `GIDP` int DEFAULT NULL,
  PRIMARY KEY (`playerID`,`yearID`,`stint`),
  CONSTRAINT `batting_people_playerID_fk` FOREIGN KEY (`playerID`) REFERENCES `people` (`playerID`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci
~~~~

# Part D: SQL Queries

NOTE: You must use the CLEAN lahman schema provided in HW0 for the queries below to ensure your answers are consistent with the solutions. 

In [11]:
%reload_ext sql
%sql mysql+pymysql://root:19990105Hyy@localhost/lahmansbaseballdb

## Question 0
What is the average salary in baseball history?

In [12]:
%sql select avg(salary)from salaries

 * mysql+pymysql://root:***@localhost/lahmansbaseballdb
   mysql+pymysql://root:***@localhost/lahmansdb_to_clean
1 rows affected.


[(2085634.053125473,)]

## Question 1
Select the players with a first name of Sam who were born in the United States and attended college.

Include their first name, last name, playerID, school ID, yearID and birth state. Limit 10

Hint: Use a Join between People and CollegePlaying

In [13]:
%%sql 
select 
    nameFirst,nameLast,people.playerID,schoolID,yearID,birthState
from
    people,collegeplaying
where
    people.birthCountry = 'USA'
    and people.nameFirst = 'Sam'
    and people.playerID = collegeplaying.playerID
limit 10


 * mysql+pymysql://root:***@localhost/lahmansbaseballdb
   mysql+pymysql://root:***@localhost/lahmansdb_to_clean
10 rows affected.


[('Sam', 'Barnes', 'barnesa01', 'auburn', 1918, 'AL'),
 ('Sam', 'Barnes', 'barnesa01', 'auburn', 1919, 'AL'),
 ('Sam', 'Barnes', 'barnesa01', 'auburn', 1920, 'AL'),
 ('Sam', 'Barnes', 'barnesa01', 'auburn', 1921, 'AL'),
 ('Sam', 'Bowens', 'bowensa01', 'tennst', 1956, 'NC'),
 ('Sam', 'Bowens', 'bowensa01', 'tennst', 1957, 'NC'),
 ('Sam', 'Bowens', 'bowensa01', 'tennst', 1958, 'NC'),
 ('Sam', 'Bowen', 'bowensa02', 'gacoast', 1971, 'GA'),
 ('Sam', 'Bowen', 'bowensa02', 'gacoast', 1972, 'GA'),
 ('Sam', 'Brown', 'brownsa01', 'grovecity', 1899, 'PA')]

## Question 2
Update all entries with full_name Columbia University to 'Columbia University in the City of New York' in the Schools table. Then select the row.


In [14]:
%%sql
update schools set name_full = 'Columbia University in the City of New York'
where name_full = 'Columbia University';
select *
from schools
where name_full = 'Columbia University in the City of New York';

 * mysql+pymysql://root:***@localhost/lahmansbaseballdb
   mysql+pymysql://root:***@localhost/lahmansdb_to_clean
0 rows affected.
1 rows affected.


[('columbia', 'Columbia University in the City of New York', 'New York', 'NY', 'USA')]

# Part E: CSVDataTable


## i. Conceptual Questions
The purpose of this homework is to teach you the behaviour of SQL Databases by asking you to implement functions that will model the behaviour of a real database with CSVDataTable. You will mimic a SQL Database using CSV files. 

Read through the scaffolding code provided in the CSVDataTable folder first to understand and answer the following conceptual questions.

1. Given this SQL statement:
    
        SELECT nameFirst, nameLast FROM people WHERE playerID = collied01
   
   If you run find_by_primary_key() on this statement, what are key_fields and field_list? 

    <i> key_fields: collied01  \
    field_list: nameFirst, nameLast </i>

2. What should be checked when you are trying to INSERT a new row into a table with a PK?

    <i> The primary key playerID is unique. </i>

3. What should be checked when you are trying to UPDATE a row in a table with a PK?

    <i> The primary key playerID is unique. </i>

## ii. Coding


You are responsible for implementing and testing two classes in Python: CSVDataTable, BaseDataTable.
The python files and data can be found in the assignment under Courseworks. 

We have already given you **find_by_template(self, template, field_list=None, limit=None, offset=None, order_by=None)**
Use this as a jumping off point for the rest of your functions.

Methods to complete:

CSVDataTable.py
- find_by_primary_key(self, key_fields, field_list=None)
- delete_by_key(self, key_fields)
- delete_by_template(self, template)
- update_by_key(self, key_fields, new_values)
- update_by_template(self, template, new_values)
- insert(self, new_record)
CSV_table_tests.py
- You must test all methods. You will have to write these tests yourself. 
- You must test your methods on the People and Batting table.

If you do not include tests and tests outputs 50% of this section's points will be deducted at the start




## iii. Testing

Please copy the text from the output of your tests and paste it below:


~~~
"/Users/yanyihuo/Documents/columbia/COMS W4111 Intro to database/venv/bin/python" "/Users/yanyihuo/Documents/columbia/COMS W4111 Intro to database/HW1/HW1/tests/csv_table_tests.py"
DEBUG:root:CSVDataTable.__init__: data = {
  "table_name": "People",
  "connect_info": {
    "directory": "/Users/yanyihuo/Documents/columbia/COMS W4111 Intro to database/HW1/HW1/Data/Baseball",
    "file_name": "People.csv"
  },
  "key_columns": [
    "playerID"
  ],
  "debug": true
}

find_by_primary_key(): Known Record
{'playerID': 'aardsda01', 'birthYear': '1981', 'birthMonth': '12', 'birthDay': '27', 'birthCountry': 'USA', 'birthState': 'CO', 'birthCity': 'Denver', 'deathYear': '', 'deathMonth': '', 'deathDay': '', 'deathCountry': '', 'deathState': '', 'deathCity': '', 'nameFirst': 'David', 'nameLast': 'Aardsma', 'nameGiven': 'David Allan', 'weight': '215', 'height': '75', 'bats': 'R', 'throws': 'R', 'debut': '2004-04-06', 'finalGame': '2015-08-23', 'retroID': 'aardd001', 'bbrefID': 'aardsda01'}

find_by_primary_key(): Known Record
{'playerID': 'barnebo01', 'birthYear': '1902', 'birthMonth': '1', 'birthDay': '6', 'birthCountry': 'USA', 'birthState': 'IL', 'birthCity': 'Washburn', 'deathYear': '1993', 'deathMonth': '12', 'deathDay': '8', 'deathCountry': 'USA', 'deathState': 'IL', 'deathCity': 'Peoria', 'nameFirst': 'Bob', 'nameLast': 'Barnes', 'nameGiven': 'Robert Avery', 'weight': '150', 'height': '71', 'bats': 'L', 'throws': 'L', 'debut': '1924-07-08', 'finalGame': '1924-07-10', 'retroID': 'barnb104', 'bbrefID': 'barnebo01'}

find_by_primary_key(): Unknown Record
DEBUG:root:CSVDataTable._load: Loaded 19619 rows
None

find_by_template(): Known Template
[{'playerID': 'aardsda01', 'birthYear': '1981', 'birthMonth': '12', 'birthDay': '27', 'birthCountry': 'USA', 'birthState': 'CO', 'birthCity': 'Denver', 'deathYear': '', 'deathMonth': '', 'deathDay': '', 'deathCountry': '', 'deathState': '', 'deathCity': '', 'nameFirst': 'David', 'nameLast': 'Aardsma', 'nameGiven': 'David Allan', 'weight': '215', 'height': '75', 'bats': 'R', 'throws': 'R', 'debut': '2004-04-06', 'finalGame': '2015-08-23', 'retroID': 'aardd001', 'bbrefID': 'aardsda01'}]

find_by_template(): Unknown Template
[]

delete_by_key(): Known record
1
find the deleted key:
None

delete_by_key(): Unknown record
0

delete_by_template(): Known Template
64
find the deleted record:
[]

delete_by_template(): Unknown Template
0

update_by_key(): does not update key
1
{'playerID': 'barnebo01', 'birthYear': '2000', 'birthMonth': '1', 'birthDay': '6', 'birthCountry': 'USA', 'birthState': 'IL', 'birthCity': 'Washburn', 'deathYear': '1993', 'deathMonth': '12', 'deathDay': '8', 'deathCountry': 'USA', 'deathState': 'IL', 'deathCity': 'Peoria', 'nameFirst': 'Bob', 'nameLast': 'Barnes', 'nameGiven': 'Robert Avery', 'weight': '150', 'height': '71', 'bats': 'L', 'throws': 'L', 'debut': '1924-07-08', 'finalGame': '1924-07-10', 'retroID': 'barnb104', 'bbrefID': 'barnebo01'}

update_by_key():  update key and key is duplicate
0
{'playerID': 'aaronto01', 'birthYear': '1939', 'birthMonth': '8', 'birthDay': '5', 'birthCountry': 'USA', 'birthState': 'AL', 'birthCity': 'Mobile', 'deathYear': '1984', 'deathMonth': '8', 'deathDay': '16', 'deathCountry': 'USA', 'deathState': 'GA', 'deathCity': 'Atlanta', 'nameFirst': 'Tommie', 'nameLast': 'Aaron', 'nameGiven': 'Tommie Lee', 'weight': '190', 'height': '75', 'bats': 'R', 'throws': 'R', 'debut': '1962-04-10', 'finalGame': '1971-09-26', 'retroID': 'aarot101', 'bbrefID': 'aaronto01'}

update_by_key():  update key but key is unique
1
{'playerID': 'aaaaa', 'birthYear': '2000', 'birthMonth': '1', 'birthDay': '6', 'birthCountry': 'USA', 'birthState': 'IL', 'birthCity': 'Washburn', 'deathYear': '1993', 'deathMonth': '12', 'deathDay': '8', 'deathCountry': 'USA', 'deathState': 'IL', 'deathCity': 'Peoria', 'nameFirst': 'Bob', 'nameLast': 'Barnes', 'nameGiven': 'Robert Avery', 'weight': '150', 'height': '71', 'bats': 'L', 'throws': 'L', 'debut': '1924-07-08', 'finalGame': '1924-07-10', 'retroID': 'barnb104', 'bbrefID': 'barnebo01'}

update_by_template(): does not update key
59

update_by_template(): update key and key is duplicate
0

update_by_template():  update key but key is unique
1

insert(): unique record
insert successfully
None

insert(): duplicate record
insert unsuccessfully
None
DEBUG:root:CSVDataTable.__init__: data = {
  "table_name": "Batting",
  "connect_info": {
    "directory": "/Users/yanyihuo/Documents/columbia/COMS W4111 Intro to database/HW1/HW1/Data/Baseball",
    "file_name": "Batting.csv"
  },
  "key_columns": [
    "playerID",
    "yearID",
    "stint"
  ],
  "debug": true
}

find_by_primary_key(): Known Record
DEBUG:root:CSVDataTable._load: Loaded 105861 rows
{'playerID': 'abercda01', 'yearID': '1871', 'stint': '1', 'teamID': 'TRO', 'lgID': 'NA', 'G': '1', 'AB': '4', 'R': '0', 'H': '0', '2B': '0', '3B': '0', 'HR': '0', 'RBI': '0', 'SB': '0', 'CS': '0', 'BB': '0', 'SO': '0', 'IBB': '', 'HBP': '', 'SH': '', 'SF': '', 'GIDP': '0'}

find_by_primary_key(): Unknown Record
None

find_by_template(): Known Template
[{'playerID': 'abercda01', 'yearID': '1871', 'stint': '1', 'teamID': 'TRO', 'lgID': 'NA', 'G': '1', 'AB': '4', 'R': '0', 'H': '0', '2B': '0', '3B': '0', 'HR': '0', 'RBI': '0', 'SB': '0', 'CS': '0', 'BB': '0', 'SO': '0', 'IBB': '', 'HBP': '', 'SH': '', 'SF': '', 'GIDP': '0'}]

find_by_template(): Unknown Template
[]

delete_by_key(): Known record
1
find the deleted key:
None

delete_by_key(): Unknown record
0

delete_by_template(): Known Template
23
find the deleted record:
[]

delete_by_template(): Unknown Template
0

update_by_key(): does not update key
1
{'playerID': 'addybo01', 'yearID': '1871', 'stint': '1', 'teamID': 'HHH', 'lgID': 'NA', 'G': '25', 'AB': '118', 'R': '30', 'H': '32', '2B': '6', '3B': '0', 'HR': '0', 'RBI': '13', 'SB': '8', 'CS': '1', 'BB': '4', 'SO': '0', 'IBB': '', 'HBP': '', 'SH': '', 'SF': '', 'GIDP': '0'}

update_by_key():  update key and key is duplicate
0
{'playerID': 'addybo01', 'yearID': '1871', 'stint': '1', 'teamID': 'HHH', 'lgID': 'NA', 'G': '25', 'AB': '118', 'R': '30', 'H': '32', '2B': '6', '3B': '0', 'HR': '0', 'RBI': '13', 'SB': '8', 'CS': '1', 'BB': '4', 'SO': '0', 'IBB': '', 'HBP': '', 'SH': '', 'SF': '', 'GIDP': '0'}

update_by_key():  update key but key is unique
1
{'playerID': 'aaaaa', 'yearID': '1871', 'stint': '1', 'teamID': 'HHH', 'lgID': 'NA', 'G': '25', 'AB': '118', 'R': '30', 'H': '32', '2B': '6', '3B': '0', 'HR': '0', 'RBI': '13', 'SB': '8', 'CS': '1', 'BB': '4', 'SO': '0', 'IBB': '', 'HBP': '', 'SH': '', 'SF': '', 'GIDP': '0'}

update_by_template(): does not update key
25

update_by_template(): update key and key is duplicate
0

update_by_template():  update key but key is unique
2

insert(): unique record
insert successfully
None

insert(): unique record
insert successfully
None

insert(): duplicate record
insert unsuccessfully
None

Process finished with exit code 0

~~~